### PR TITLE
Fix teardown deadlock between receive close and send close

### PR DESCRIPTION
**Pull request by tvegas1**
## What
Fix deadlock on teardown path at scale.

## Why
Call for `nccl_ucx_free_worker()` remains pending on `ncclSocketRecv()` for remote close signal.

## How
Wait for all close procedures to be completed (send and recv) before actually pulling remote pending signals.

### CODE REVIEW

**Comment by tvegas1** [comment]
pls look @brminich

File: src/ucx_plugin.c [hunk 1]
```diff
@@ -297,46 +301,60 @@ static ncclResult_t ucx_get_ctx_and_worker(int dev, ucp_context_h *ctx,
...
   pthread_mutex_unlock(&nccl_ucx_lock);
 
-  if (i < ncclNIbDevs && count == 0) {
-    ep = workers[i].eps;
-    while(ep){
+  if (!done) {
+    return ncclSuccess;
+  }
+
+  for (ucx_worker = worker_gc_list; ucx_worker != NULL; ucx_worker = next) {
```
Comment (brminich) [inline src/ucx_plugin.c:333]: do we really need `gc_list` or we can just iterate over device worker once free was invoked for all of them?

Comment (tvegas1) [reply src/ucx_plugin.c:333]: agree, we could iterate over the `workers[]` but the plan is to remove that array altogether and allocate workers maybe rx/tx based + max number of ep per worker combination.

Comment (brminich) [reply src/ucx_plugin.c:333]: ok

**Review by bureddy — APPROVED** [review]

**Pull request merged**